README
doc/ruid.mmd

Changes
Revision history for PostgreSQL extension semver.

0.0.4 
      - Adjust abstracts.
      - Rephrase some sentences in documentation.
      - Provide details about encoding and padding.
      - Add helper functions.
      - Describe internals a bit.

0.0.3 2013-02-02 15:41:25
      - Zap old and commented statements.
      - Write about future plans.

0.0.2 2013-02-02 15:19:26
      - Documentation fixes.
      - Add Changes.

0.0.1 2013-02-02 15:08:00
      - Initial release.


src/ruid.c
/*
 * Copyright (c) 2010-2013 Simon Bertrang <janus@errornet.de>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "postgres.h"

#include "access/hash.h"
#include "libpq/pqformat.h"
#include "utils/builtins.h"

PG_MODULE_MAGIC;

int	b64_pton(char const *, u_char *, size_t);
int	b64_ntop(u_char const *, size_t, char *, size_t);

/* ruid size in bytes */
#define RUID_LEN 16

/* pg_ruid_t is declared to be struct pg_uuid_t in uuid.h */
typedef struct pg_ruid_t
{
	unsigned char data[RUID_LEN];
} pg_ruid_t;

#define DatumGetRUIDP(X)	((pg_ruid_t *)DatumGetPointer(X))
#define PG_GETARG_RUID_P(X)	DatumGetRUIDP(PG_GETARG_DATUM(X))

#define RUIDPGetDatum(X)	PointerGetDatum(X)
#define PG_RETURN_RUID_P(X)	return RUIDPGetDatum(X)

static void	string_to_ruid(const char *, pg_ruid_t *);
static int	ruid_internal_cmp(const pg_ruid_t *, const pg_ruid_t *);

Datum ruid_in(PG_FUNCTION_ARGS);
Datum ruid_out(PG_FUNCTION_ARGS);
Datum ruid_recv(PG_FUNCTION_ARGS);
Datum ruid_send(PG_FUNCTION_ARGS);
Datum ruid_lt(PG_FUNCTION_ARGS);
Datum ruid_le(PG_FUNCTION_ARGS);
Datum ruid_eq(PG_FUNCTION_ARGS);
Datum ruid_ge(PG_FUNCTION_ARGS);
Datum ruid_gt(PG_FUNCTION_ARGS);
Datum ruid_ne(PG_FUNCTION_ARGS);
Datum ruid_cmp(PG_FUNCTION_ARGS);
Datum ruid_hash(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(ruid_in);

Datum
ruid_in(PG_FUNCTION_ARGS)
{
	pg_ruid_t  *ruid;
	char	   *ruid_str = PG_GETARG_CSTRING(0);

	ruid = (pg_ruid_t *)palloc(sizeof(*ruid));

	string_to_ruid(ruid_str, ruid);

	PG_RETURN_RUID_P(ruid);
}

PG_FUNCTION_INFO_V1(ruid_out);

Datum
ruid_out(PG_FUNCTION_ARGS)
{
	StringInfoData	 buf;
	unsigned char	 in[18];
	unsigned int	 n = 0;
	pg_ruid_t	*ruid = PG_GETARG_RUID_P(0);
	char		 out[25];

	bzero(&out, sizeof(out));

	in[n] = '\0';
	while (++n < sizeof(in)-1)
		in[n] = ruid->data[n-1];
	in[n++] = '\0';


	/* encode uuid data in base64 */
	n = b64_ntop(in, sizeof(in), out, sizeof(out));

	initStringInfo(&buf);

	/* start before padding of the base64 string, walk backwards and
	 * replace special chars
	 */
	for (n = 1; n < sizeof(out)-2; n++)
		if (out[n] == '+')
			appendStringInfoChar(&buf, '-');
		else if (out[n] == '/')
			appendStringInfoChar(&buf, '_');
		else
			appendStringInfoChar(&buf, out[n]);

	PG_RETURN_CSTRING(buf.data);
}

/*
 * We allow RUIDs as a series of 32 hexadecimal digits with an optional dash
 * after each group of 4 hexadecimal digits, and optionally surrounded by {}.
 * (The canonical format 8x-4x-4x-4x-12x, where "nx" means n hexadecimal
 * digits, is the only one used for output.)
 */
static void
string_to_ruid(const char *string, pg_ruid_t *ruid)
{
	const char	*src = string;
	size_t		 len = strlen(string);
	bool		 braces = false;
	int		 i;

	/* get and check RUID length */
	if (len == 22) {
		unsigned char	 out[18];
		unsigned int	 n = 0;
		char		 buf[25];

		buf[n] = 'A';
		while (++n < sizeof(buf)-2)
			switch (src[n-1]) {
			case '-': buf[n] = '+'; break;
			case '_': buf[n] = '/'; break;
			default:
				if (!(isalpha(src[n-1]) || isdigit(src[n-1])))
					goto syntax_error;
				buf[n] = src[n-1];
			}
		buf[n++] = 'A';
		buf[n++] = '\0';

		bzero(&out, sizeof(out));

		if (b64_pton(buf, out, sizeof(out)) != sizeof(out))
			goto syntax_error;

		/* ignore the leading padding byte */
		memcpy(ruid->data, out+1, sizeof(ruid->data));

		return;
	}

	/* otherwise check for uuid syntax */

	if (src[0] == '{') {
		src++;
		braces = true;
	}

	for (i = 0; i < RUID_LEN; i++) {
		char		str_buf[3];

		if (src[0] == '\0' || src[1] == '\0')
			goto syntax_error;
		memcpy(str_buf, src, 2);
		if (!isxdigit((unsigned char) str_buf[0]) ||
			!isxdigit((unsigned char) str_buf[1]))
			goto syntax_error;

		str_buf[2] = '\0';
		ruid->data[i] = (unsigned char) strtoul(str_buf, NULL, 16);
		src += 2;
		if (src[0] == '-' && (i % 2) == 1 && i < RUID_LEN - 1)
			src++;
	}

	if (braces) {
		if (*src != '}')
			goto syntax_error;
		src++;
	}

	if (*src != '\0')
		goto syntax_error;

	return;

syntax_error:
	ereport(ERROR,
			(errcode(ERRCODE_INVALID_TEXT_REPRESENTATION),
			 errmsg("invalid input syntax for ruid: \"%s\"",
					string)));
}

PG_FUNCTION_INFO_V1(ruid_recv);

Datum
ruid_recv(PG_FUNCTION_ARGS)
{
	StringInfo	 buffer = (StringInfo)PG_GETARG_POINTER(0);
	pg_ruid_t	*ruid;

	ruid = (pg_ruid_t *)palloc(RUID_LEN);

	memcpy(ruid->data, pq_getmsgbytes(buffer, RUID_LEN), RUID_LEN);

	PG_RETURN_POINTER(ruid);
}

PG_FUNCTION_INFO_V1(ruid_send);

Datum
ruid_send(PG_FUNCTION_ARGS)
{
	StringInfoData	 buffer;
	pg_ruid_t	*ruid = PG_GETARG_RUID_P(0);

	pq_begintypsend(&buffer);
	pq_sendbytes(&buffer, (char *)ruid->data, RUID_LEN);

	PG_RETURN_BYTEA_P(pq_endtypsend(&buffer));
}

/* internal ruid compare function */
static int
ruid_internal_cmp(const pg_ruid_t *arg1, const pg_ruid_t *arg2)
{
	return memcmp(arg1->data, arg2->data, RUID_LEN);
}

PG_FUNCTION_INFO_V1(ruid_lt);

Datum
ruid_lt(PG_FUNCTION_ARGS)
{
	pg_ruid_t	*arg1 = PG_GETARG_RUID_P(0);
	pg_ruid_t	*arg2 = PG_GETARG_RUID_P(1);

	PG_RETURN_BOOL(ruid_internal_cmp(arg1, arg2) < 0);
}

PG_FUNCTION_INFO_V1(ruid_le);

Datum
ruid_le(PG_FUNCTION_ARGS)
{
	pg_ruid_t	*arg1 = PG_GETARG_RUID_P(0);
	pg_ruid_t	*arg2 = PG_GETARG_RUID_P(1);

	PG_RETURN_BOOL(ruid_internal_cmp(arg1, arg2) <= 0);
}

PG_FUNCTION_INFO_V1(ruid_eq);

Datum
ruid_eq(PG_FUNCTION_ARGS)
{
	pg_ruid_t	*arg1 = PG_GETARG_RUID_P(0);
	pg_ruid_t	*arg2 = PG_GETARG_RUID_P(1);

	PG_RETURN_BOOL(ruid_internal_cmp(arg1, arg2) == 0);
}

PG_FUNCTION_INFO_V1(ruid_ge);

Datum
ruid_ge(PG_FUNCTION_ARGS)
{
	pg_ruid_t	*arg1 = PG_GETARG_RUID_P(0);
	pg_ruid_t	*arg2 = PG_GETARG_RUID_P(1);

	PG_RETURN_BOOL(ruid_internal_cmp(arg1, arg2) >= 0);
}

PG_FUNCTION_INFO_V1(ruid_gt);

Datum
ruid_gt(PG_FUNCTION_ARGS)
{
	pg_ruid_t	*arg1 = PG_GETARG_RUID_P(0);
	pg_ruid_t	*arg2 = PG_GETARG_RUID_P(1);

	PG_RETURN_BOOL(ruid_internal_cmp(arg1, arg2) > 0);
}

PG_FUNCTION_INFO_V1(ruid_ne);

Datum
ruid_ne(PG_FUNCTION_ARGS)
{
	pg_ruid_t	*arg1 = PG_GETARG_RUID_P(0);
	pg_ruid_t	*arg2 = PG_GETARG_RUID_P(1);

	PG_RETURN_BOOL(ruid_internal_cmp(arg1, arg2) != 0);
}

PG_FUNCTION_INFO_V1(ruid_cmp);

/* handler for btree index operator */
Datum
ruid_cmp(PG_FUNCTION_ARGS)
{
	pg_ruid_t	*arg1 = PG_GETARG_RUID_P(0);
	pg_ruid_t	*arg2 = PG_GETARG_RUID_P(1);

	PG_RETURN_INT32(ruid_internal_cmp(arg1, arg2));
}

PG_FUNCTION_INFO_V1(ruid_hash);

/* hash index support */
Datum
ruid_hash(PG_FUNCTION_ARGS)
{
	pg_ruid_t	*key = PG_GETARG_RUID_P(0);

	PG_RETURN_INT32(hash_any(key->data, RUID_LEN));
}



src/base64.c
/*	$OpenBSD: base64.c,v 1.5 2006/10/21 09:55:03 otto Exp $	*/

/*
 * Copyright (c) 1996 by Internet Software Consortium.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM DISCLAIMS
 * ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL INTERNET SOFTWARE
 * CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 * PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
 * ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
 * SOFTWARE.
 */

/*
 * Portions Copyright (c) 1995 by International Business Machines, Inc.
 *
 * International Business Machines, Inc. (hereinafter called IBM) grants
 * permission under its copyrights to use, copy, modify, and distribute this
 * Software with or without fee, provided that the above copyright notice and
 * all paragraphs of this notice appear in all copies, and that the name of IBM
 * not be used in connection with the marketing of any product incorporating
 * the Software or modifications thereof, without specific, written prior
 * permission.
 *
 * To the extent it has a right to do so, IBM grants an immunity from suit
 * under its patents, if any, for the use, sale or manufacture of products to
 * the extent that such products are used for performing Domain Name System
 * dynamic updates in TCP/IP networks by means of the Software.  No immunity is
 * granted for any product per se or for any other function of any product.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", AND IBM DISCLAIMS ALL WARRANTIES,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE.  IN NO EVENT SHALL IBM BE LIABLE FOR ANY SPECIAL,
 * DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE, EVEN
 * IF IBM IS APPRISED OF THE POSSIBILITY OF SUCH DAMAGES.
 */

#include <sys/types.h>
#include <ctype.h>
#include <string.h>

int	b64_pton(char const *, u_char *, size_t);
int	b64_ntop(u_char const *, size_t, char *, size_t);



static const char Base64[] =
	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
static const char Pad64 = '=';

/* (From RFC1521 and draft-ietf-dnssec-secext-03.txt)
   The following encoding technique is taken from RFC 1521 by Borenstein
   and Freed.  It is reproduced here in a slightly edited form for
   convenience.

   A 65-character subset of US-ASCII is used, enabling 6 bits to be
   represented per printable character. (The extra 65th character, "=",
   is used to signify a special processing function.)

   The encoding process represents 24-bit groups of input bits as output
   strings of 4 encoded characters. Proceeding from left to right, a
   24-bit input group is formed by concatenating 3 8-bit input groups.
   These 24 bits are then treated as 4 concatenated 6-bit groups, each
   of which is translated into a single digit in the base64 alphabet.

   Each 6-bit group is used as an index into an array of 64 printable
   characters. The character referenced by the index is placed in the
   output string.

                         Table 1: The Base64 Alphabet

      Value Encoding  Value Encoding  Value Encoding  Value Encoding
          0 A            17 R            34 i            51 z
          1 B            18 S            35 j            52 0
          2 C            19 T            36 k            53 1
          3 D            20 U            37 l            54 2
          4 E            21 V            38 m            55 3
          5 F            22 W            39 n            56 4
          6 G            23 X            40 o            57 5
          7 H            24 Y            41 p            58 6
          8 I            25 Z            42 q            59 7
          9 J            26 a            43 r            60 8
         10 K            27 b            44 s            61 9
         11 L            28 c            45 t            62 +
         12 M            29 d            46 u            63 /
         13 N            30 e            47 v
         14 O            31 f            48 w         (pad) =
         15 P            32 g            49 x
         16 Q            33 h            50 y

   Special processing is performed if fewer than 24 bits are available
   at the end of the data being encoded.  A full encoding quantum is
   always completed at the end of a quantity.  When fewer than 24 input
   bits are available in an input group, zero bits are added (on the
   right) to form an integral number of 6-bit groups.  Padding at the
   end of the data is performed using the '=' character.

   Since all base64 input is an integral number of octets, only the
         -------------------------------------------------                       
   following cases can arise:
   
       (1) the final quantum of encoding input is an integral
           multiple of 24 bits; here, the final unit of encoded
	   output will be an integral multiple of 4 characters
	   with no "=" padding,
       (2) the final quantum of encoding input is exactly 8 bits;
           here, the final unit of encoded output will be two
	   characters followed by two "=" padding characters, or
       (3) the final quantum of encoding input is exactly 16 bits;
           here, the final unit of encoded output will be three
	   characters followed by one "=" padding character.
   */

int
b64_ntop(u_char const *src, size_t srclength, char *target, size_t targsize)
{
	size_t datalength = 0;
	u_char input[3];
	u_char output[4];
	int i;

	while (2 < srclength) {
		input[0] = *src++;
		input[1] = *src++;
		input[2] = *src++;
		srclength -= 3;

		output[0] = input[0] >> 2;
		output[1] = ((input[0] & 0x03) << 4) + (input[1] >> 4);
		output[2] = ((input[1] & 0x0f) << 2) + (input[2] >> 6);
		output[3] = input[2] & 0x3f;

		if (datalength + 4 > targsize)
			return (-1);
		target[datalength++] = Base64[output[0]];
		target[datalength++] = Base64[output[1]];
		target[datalength++] = Base64[output[2]];
		target[datalength++] = Base64[output[3]];
	}
    
	/* Now we worry about padding. */
	if (0 != srclength) {
		/* Get what's left. */
		input[0] = input[1] = input[2] = '\0';
		for (i = 0; i < srclength; i++)
			input[i] = *src++;
	
		output[0] = input[0] >> 2;
		output[1] = ((input[0] & 0x03) << 4) + (input[1] >> 4);
		output[2] = ((input[1] & 0x0f) << 2) + (input[2] >> 6);

		if (datalength + 4 > targsize)
			return (-1);
		target[datalength++] = Base64[output[0]];
		target[datalength++] = Base64[output[1]];
		if (srclength == 1)
			target[datalength++] = Pad64;
		else
			target[datalength++] = Base64[output[2]];
		target[datalength++] = Pad64;
	}
	if (datalength >= targsize)
		return (-1);
	target[datalength] = '\0';	/* Returned value doesn't count \0. */
	return (datalength);
}

/* skips all whitespace anywhere.
   converts characters, four at a time, starting at (or after)
   src from base - 64 numbers into three 8 bit bytes in the target area.
   it returns the number of data bytes stored at the target, or -1 on error.
 */

int
b64_pton(char const *src, u_char *target, size_t targsize)
{
	int tarindex, state, ch;
	char *pos;

	state = 0;
	tarindex = 0;

	while ((ch = *src++) != '\0') {
		if (isspace(ch))	/* Skip whitespace anywhere. */
			continue;

		if (ch == Pad64)
			break;

		pos = strchr(Base64, ch);
		if (pos == 0) 		/* A non-base64 character. */
			return (-1);

		switch (state) {
		case 0:
			if (target) {
				if (tarindex >= targsize)
					return (-1);
				target[tarindex] = (pos - Base64) << 2;
			}
			state = 1;
			break;
		case 1:
			if (target) {
				if (tarindex + 1 >= targsize)
					return (-1);
				target[tarindex]   |=  (pos - Base64) >> 4;
				target[tarindex+1]  = ((pos - Base64) & 0x0f)
							<< 4 ;
			}
			tarindex++;
			state = 2;
			break;
		case 2:
			if (target) {
				if (tarindex + 1 >= targsize)
					return (-1);
				target[tarindex]   |=  (pos - Base64) >> 2;
				target[tarindex+1]  = ((pos - Base64) & 0x03)
							<< 6;
			}
			tarindex++;
			state = 3;
			break;
		case 3:
			if (target) {
				if (tarindex >= targsize)
					return (-1);
				target[tarindex] |= (pos - Base64);
			}
			tarindex++;
			state = 0;
			break;
		}
	}

	/*
	 * We are done decoding Base-64 chars.  Let's see if we ended
	 * on a byte boundary, and/or with erroneous trailing characters.
	 */

	if (ch == Pad64) {		/* We got a pad char. */
		ch = *src++;		/* Skip it, get next. */
		switch (state) {
		case 0:		/* Invalid = in first position */
		case 1:		/* Invalid = in second position */
			return (-1);

		case 2:		/* Valid, means one byte of info */
			/* Skip any number of spaces. */
			for (; ch != '\0'; ch = *src++)
				if (!isspace(ch))
					break;
			/* Make sure there is another trailing = sign. */
			if (ch != Pad64)
				return (-1);
			ch = *src++;		/* Skip the = */
			/* Fall through to "single trailing =" case. */
			/* FALLTHROUGH */

		case 3:		/* Valid, means two bytes of info */
			/*
			 * We know this char is an =.  Is there anything but
			 * whitespace after it?
			 */
			for (; ch != '\0'; ch = *src++)
				if (!isspace(ch))
					return (-1);

			/*
			 * Now make sure for cases 2 and 3 that the "extra"
			 * bits that slopped past the last full byte were
			 * zeros.  If we don't check them, they become a
			 * subliminal channel.
			 */
			if (target && target[tarindex] != 0)
				return (-1);
		}
	} else {
		/*
		 * We ended by seeing the end of the string.  Make sure we
		 * have no partial bytes lying around.
		 */
		if (state != 0)
			return (-1);
	}

	return (tarindex);
}
